### PR TITLE
Add unit-test for UpwardFileSearch().

### PR DESCRIPTION
**Pull request by hzeller**
Precondition to change the implementation to use
std::filesystem.

Issues: #765 #307

Signed-off-by: Henner Zeller <h.zeller@acm.org>

### CODE REVIEW

File: common/util/file_util.cc [hunk 1]
```diff
@@ -174,7 +174,18 @@ absl::Status SetContents(absl::string_view filename,
 }
 
 std::string JoinPath(absl::string_view base, absl::string_view name) {
-  return absl::StrCat(base, "/", name);
+  // TODO: this will certainly be different on Windows
```
Comment (tgorochowik) [inline common/util/file_util.cc:177]: I assume the idea is not to worry about that for now as it will be fixed when switching to c++17?

File: common/util/file_util.cc [hunk 2]
```diff
@@ -174,7 +174,18 @@ absl::Status SetContents(absl::string_view filename,
...
+  static constexpr absl::string_view kFileSeparator = "/";
+
+  // Make sure that we don't concatenation multiple superfluous separators.
+  // Note, since we're using string_view, the substr() operations are cheap.
+  while (!base.empty() && base[base.length() - 1] == kFileSeparator[0]) {
+    base = base.substr(0, base.length() - 1);
+  }
```
Comment (tgorochowik) [inline common/util/file_util.cc:184]: It is not immediately clear that this will work if base is just `"/"` - but I guess it will as the separator will be the leading `/` then - maybe we could add a comment saying why it's okay?

File: common/util/file_util.cc [hunk 3]
```diff
@@ -174,7 +174,18 @@ absl::Status SetContents(absl::string_view filename,
 }
 
 std::string JoinPath(absl::string_view base, absl::string_view name) {
-  return absl::StrCat(base, "/", name);
+  // TODO: this will certainly be different on Windows
+  static constexpr absl::string_view kFileSeparator = "/";
+
+  // Make sure that we don't concatenation multiple superfluous separators.
```
Comment (tgorochowik) [inline common/util/file_util.cc:180]: `concatenate`

**Review by tgorochowik — APPROVED** [review]
LGTM! Thank you for working on this!

File: common/util/file_util_test.cc [hunk 4]
```diff
@@ -61,6 +64,18 @@ TEST(FileUtil, Stem) {
   EXPECT_EQ(file::Stem("/foo/bar"), "/foo/bar");
 }
 
+TEST(FileUtil, JoinPath) {
+  EXPECT_EQ(file::JoinPath("foo", ""), "foo/");
+  EXPECT_EQ(file::JoinPath("", "bar"), "/bar");
```
Comment (tgorochowik) [inline common/util/file_util_test.cc:69]: Could we also add
`EXPECT_EQ(file::JoinPath("/", "bar"), "/bar");`

File: common/util/file_util.cc (hunk 1 above)
Comment (hzeller) [reply common/util/file_util.cc:177]: Yes. I'll work on using `std::filesystem` soon, so this is mostly a reminder.

File: common/util/file_util.cc (hunk 3 above)
Comment (hzeller) [reply common/util/file_util.cc]: Done.

File: common/util/file_util.cc (hunk 2 above)
Comment (hzeller) [reply common/util/file_util.cc:184]: I solved it by adding a test to make sure that this corner-case is covered, since anyone wondering about the implementation will look there first.

File: common/util/file_util_test.cc (hunk 4 above)
Comment (hzeller) [reply common/util/file_util_test.cc:69]: Done, also with a few more variants.
